lib/record_store/provider/ns1/patch_api_header.rb

```diff
@@ -18,15 +25,17 @@ def process_response(response)
         rate_limit.wait(sleep_time)
       end
 
-      body = JSON.parse(response.body)
-      case response
-      when Net::HTTPOK
-        NS1::Response::Success.new(body, response.code.to_i)
-      else
-        NS1::Response::Error.new(body, response.code.to_i)
+      begin
+        body = JSON.parse(response.body)
+        case response
```

Comment:
Totally optional because this works:

What do you think of extending this switch/case statement to handle the `Net::HTTPServiceUnavailable` status code, and raise the same `UnparsableBodyError`?

Just noticing [here](https://github.com/Shopify/record_store/pull/181/files#diff-5f6190009b4d6922fcd1b3bf4862de14R157-R161) that the debug messages indicate that we're retrying because the service is unavailable, yet the error we're handling is unparseable JSON.

Perhaps `UnparsableBodyError` is really `ServiceUnavailableError`?

Reply:
I'm the one that suggested moving to `UnparsableBodyError` since it allows to consider something is wrong, but that the error code is not necessarily `ServiceUnavailable`. This avoids requiring to handle every little status code that would leave us with a body that's not JSON, and handle the fact that if the body is not parsable, we will wait and retry.

Maybe we want to have a clearer error message when it happens though. Would require create a separate waiter though, or simply update the waiter class's `wait` method to take an optional `message` parameter that, when provided, goes to replace the default wait message. That way we could simply call `waiter.wait(message: 'Unable to parse response body')` for instance.

Reply:
Whoops, thanks for catching the old message @ryandv 
I think `BackoffWaiter` has a setter method for `message`, so I can use that :)

Reply:
> I think `BackoffWaiter` has a setter method for `message`, so I can use that :)

It does but then you'll have to use it everywhere you use the backoff waiter! Hence my suggestion of allowing "on-use override" of that message. That being said. doing a `waiter.message = "xxxx"` just before each call to `waiter.wait` will work. Just not as neat to read IMO, but not that important!

Reply:
Missed that comment, sorry! Agreed, on-use message setting would be better.